TESTING.md

```diff
@@ -0,0 +1,293 @@
+## Testing
+
+`canvas-kit` has 3 levels of testing: unit, functional and visual. Each area has special use-cases
+even though tests at various levels may overlap. Here's a simplified explanation of what each level
+is meant to test:
+
+- **Unit**: This level is meant to test the input/output of units of functionality (e.g. utility
+  functions or input props of components that output DOM).
+- **Functional**: This level is browser-based testing and is where UI behavior and accessibility
+  requirements are specified (e.g., Given a modal example, When clicking on the button, Then the
+  modal should be open). Functional tests are against Storybook Stories.
+- **Visual**: This level is where all styling is tested. Some props only have visual effects and it
+  is difficult to reason about at any other level. All Storybook Stories are automatically visual
+  regression tested.
+
+Tests are supposed to be a source of feedback and confidence. To achieve that, a test should be
+clear about intent and failures. Developers tend to ignore passing tests and focus only on failing
+tests. If the intent of the test cannot be determined, it is difficult to fix the test. If the
+failure isn't immediately obvious, too much time is spent fixing the test. Canvas Kit adopts a BDD
+(behavior-driven development) approach of testing. The most common way to think is about behavior is
+"given/when/then". These behaviors are a series of specifications and not necessarily "tests".
+Testing in this repo is not just testing the correctness of our components, but running a series of
+specifications. Specifications tend to be more granular to give context of what code is expected to
+do. A "test" simply has to pass. A specification requires meaning.
+
+**Bad**:
+
+```tsx
+test('SomeComponent should render correctly', async () => {
+  const {getByTestId} = render(<SomeComponent data-testid="test" text="foo" />);
```

Comment:
@NicholasBoll originally after reading this I assumed it was best practice to use `data-testid` as a query since it is used throughout this document. However after looking at the RTL docs looks like this should be used as a last choice for a query: https://testing-library.com/docs/dom-testing-library/api-queries#bytestid

Should we also adopt this convention? Or am I ok using `data-testid` for majority of my selectors?

Reply:
I advise using `data-testid` as a last resort. In Cypress, component helpers can be used to help with selecting sub-components and can be selected usually by accessibility attributes. Helpers should be the only part of the test that has implementation details about how the DOM is structured. This allows updates to the Component + component helpers without changes to the test file. In the past, testing helpers is part of the CI process guaranteeing they work for downstream consumers of Canvas Kit. If done right, this means applications can be tested using Cypress without worry of implementation changes breaking tests.

In React Testing Library, we're often testing for specific attributes and not using helpers. Helpers are additional logic that should be tested to avoid confusing failure messages. Often in React Testing Library, I'm only testing implementation details, but I want to avoid flaky tests that are difficult to debug. Most of our components allow for additional attributes to be applied to container components. This allows for easy selecting. The alternative is often `container.children[0].children[1]` or something like that. A failure due to changes in DOM structure are unhelpful.

That being said, I don't recommend using `data-testid` inside the Component source code. I don't have much issue adding them only to the tests.

The guiding principals say the same thing from the link you sent:
> In the spirit of the guiding principles, it is recommended to use this only after the other queries don't work for your use case. Using data-testid attributes do not resemble how your software is used and should be avoided if possible. That said, they are way better than querying based on DOM structure or styling css class names. Learn more about data-testids from the blog post "Making your UI tests resilient to change"

You should use other selectors when available. The best are `getByLabelText` which will return the input/textarea/select/etc that is associated with the label by the label's text. It is easy to understand and assumes that all form fields are accessible (which they should be). There are other helpers that work well for data entry and accessible roles.

Kent C. Dodds seems to be saying much the same thing here: https://kentcdodds.com/blog/making-your-ui-tests-resilient-to-change

As for low-level component unit tests? Often the selectors provided by React Testing Library just don't give me what I need and I fall back to test ids.

Reply:
Perhaps we should link to the article to help clear things up

Reply:
I added a section about selectors here: https://github.com/Workday/canvas-kit/pull/382/files#diff-84889d0af78b35050da8d12673769ebaR76-R85

@sahlhoff @lychyi, what do you think?